lib/index.js

```diff
@@ -101,81 +41,248 @@ const scrollOptions = {
+         await this.shiftOut(Registers.DecodeMode, 0)
+         await this.shiftOut(Registers.ScanLimit, 7)
+         await this.shiftOut(Registers.Shutdown, on)
+         await this.displayTest(controller, false)
+         await this.clear(controller)
+       }
+     } else {
+       this.setController(controller)
+       await this.shiftOut(Registers.DecodeMode, 0)
+       await this.shiftOut(Registers.ScanLimit, 7)
+       await this.shiftOut(Registers.Shutdown, on)
+       await this.displayTest(controller, false)
+       await this.clear(controller)
+     }
    }
-
+   
    async clear (controller) {
-     setController(this, controller)
-     const matrix = emptyMatrix()
-     await this.set(controller, matrix)
+     const matrix = Max7219.emptyMatrix()
+     if (undefined === controller) {
+       for (controller = 0; controller < this.controllerCount; controller++) {
+         this.setController(controller)
+         await this.set(controller, matrix)
+       }
+     } else {
+       this.setController(controller)
+       await this.set(controller, matrix)
+     }
    }
  
    async set (controller, matrix) {
-     setController(this, controller)
+     this.setController(controller)
      if (matrix.length !== 8) {
        throw new Error('Matrix height must be exactly 8')
      }
 
-     return encodeMatrix(this, matrix, encodeByte)
+     return this.encodeMatrix(matrix)
    }
  
    async letter (controller, letter) {
-    setController(this, controller)
-    const matrix = getMatrix(this.font, letter)
-    return encodeMatrix(this, matrix)
+    this.setController(controller)
+    const matrix = Max7219.getMatrix(this.font, letter)
+    return this.encodeMatrix(matrix)
    }
 
    async setIntensity (controller, brightness) {
-     setController(this, controller)
-     if (0 > brightness > 15) {
-       throw new Error('Brightness must be between 0 and 15')
+     if (undefined === brightness) {
+       brightness = controller;
+       if (0 > brightness > 15) {
+         throw new Error('Brightness must be between 0 and 15')
+       }
+       for (controller = 0; controller < this.controllerCount; controller++) {
+         this.setController(controller)
+         await this.shiftOut(Registers.Intensity, brightness)
+       }
+     } else {
+       if (0 > brightness > 15) {
+         throw new Error('Brightness must be between 0 and 15')
+       }
+       this.setController(controller)
+       await this.shiftOut(Registers.Intensity, brightness)
+     }
+   }
+   
+   async display (text, start = 0, length = -1) {
+     const controllerCount = this.controllerCount;
+     // skip past start characters if start > 0
+     text = (start > 0) ?  text.substring(start) : text;
+     // Ensure we don't have more characters than controllers and ensure if text is too short we pad to controllerCount
+     // length with spaces.
+     length = Math.min(controllerCount, (length >= 0 && length <= text.length) ? length : text.length);
```

Comment:
same with this one

Reply:
Although I don't fully agree based on primitive types and immutability, both display() and scroll have been updated to use local variables instead of the parameters passed into the function.